std/permutation.asm

```diff
@@ -24,19 +24,53 @@ let compress_expression_array = |expr_array| fold(expr_array, 0, |acc, el| alpha
 // - rhs: An array of expressions
 let permutation = |acc, lhs_selector, lhs, rhs_selector, rhs| {
 
+    let GOLDILOCKS_PRIME = 2**64 - 2**32 + 1;
     let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = assert(modulus() > 2**100, || "This implementation assumes a large field");
+    let _ = assert(modulus() > 2**100 || modulus() == GOLDILOCKS_PRIME, || "No implementation on this small field");
+    let _ = assert(len(acc) == 2, || "Expected 2 accumulators");
 
-    let lhs_folded = lhs_selector * compress_expression_array(lhs);
-    let rhs_folded = rhs_selector * compress_expression_array(rhs);
+    // On the Goldilocks field, we evaluate the polynomial on the F_{p^2} extension field
+    // modulo the irreducible polynomial x^2 - 7.
+    let needs_extension = modulus() == GOLDILOCKS_PRIME;
+
+    // On the extension field, we'll need two field elements to represent the challenge.
+    // If we don't need an extension field, we can simply set the second component to 0,
+    // in which case the operations below effectively only operate on the first component.
+    let alpha = if needs_extension {[alpha1, alpha2]} else {[alpha1, 0]};
+    let beta = if needs_extension {[beta1, beta2]} else {[beta1, 0]};
+
+    // Define operations on extension field elements, which are array of length 2.
+    // Addition and subtraction is just performed element-wise.
+    // Multiplication works by multiplying out the two input polynomials and using the fact
+    // that x^2 == 7 (mod x^2 - 7) to get rid of the x^2 terms.
+    let add_ext = |a, b| [a[0] + b[0], a[1] - b[1]];
```

Comment:
What about introducing a fancy user-defined type for the extension field elements?